Support namespaced {Sriov,SriovIB,OVS}Networks

zeeke opened the pull request:
SriovNetwork,IbSriovNetwork,OVSNetwork object (XNetwork resources from now on) must be created in the operator's namespace and the `.Spec.NetworkNamespace` field defines where the
controller should create the NetworkAttachmentDefinition resource. This constraint can be a problem
in clusters where applications are managed by non cluster administrators.

These changes makes the `genericNetworkReconciler` to accept XNetwork resources in namespaces different than
the operator's one. In such cases, the field `.Spec.NetworkNamespace` must be empty, and a validating webhook
ensure this constraint.

Webhook validation for networks .Spec.NetworkNamespace field 

Move `[sriov] operator No SriovNetworkNodePolicy ...` test cases to its
own test file.

Implement test case to verify controller and webhook logic

zeeke commented on pkg/webhook/validate_networks.go:
```diff
@@ -0,0 +1,43 @@
+}
+
+func validateNetworkNamespace(cr controllers.NetworkCRInstance) error {
+	if cr.GetNamespace() != vars.Namespace && cr.NetworkNamespace() != "" {
```
It's not mandatory, and if the SriovNetwork.Spec.NetworkNamespace is empty, the NetworkattachmentDefinition is created in the sriov-network-operator namespace.

This might not sound very clean, but in openshift the `openshift-sriov-network-operator` namespace is listed as global:

```
$ oc get cm -n openshift-multus multus-daemon-config -o yaml | grep global
        "globalNamespaces": "default,openshift-multus,openshift-sriov-network-operator,openshift-cnv",
```

It's usefult to create one NetAttachDefinition and use it in multiple namespace

zeeke commented on controllers/sriovnetwork_controller_test.go:
```diff
@@ -342,6 +342,45 @@ var _ = Describe("SriovNetwork Controller", Ordered, func() {
+					UID:        cr.UID,
+					Name:       cr.Name,
+				}
+				Expect(netAttDef.ObjectMeta.OwnerReferences).To(ContainElement(expectedOwnerReference))
```
unfortunately no, OwnerReferences are cleaned by kubernetes controllers, which are not available in test env. :confused: 

We have to trust them to work correctly

zeeke commented on pkg/webhook/validate_networks.go:
```diff
@@ -0,0 +1,43 @@
+
+func validateNetworkNamespace(cr controllers.NetworkCRInstance) error {
+	if cr.GetNamespace() != vars.Namespace && cr.NetworkNamespace() != "" {
+		return fmt.Errorf(".Spec.NetworkNamespace field can't be specified if the resource is not in the %s namespace", vars.Namespace)
```
That's a little bit debatable: having two or more objects contending the same NetAttachDef is handled in 
- https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/897

It's the same problem, viewed from the perspective of the webhook, and we already have the same with an SriovNetwork + SriovIBNetwork pointing to the same object.

I prefer to address it in #897 or in another subsequent PR, after discussing in a community meeting. WDYT?

SchSeba replied:
works for me

zeeke commented on test/conformance/tests/test_no_policy.go:
```diff
@@ -0,0 +1,299 @@
+			})
+		})
+
+		DescribeTable("should gracefully restart quickly", func(webookEnabled bool) {
```
To leave it where it was, I have to put back all the `No SriovNetworkPolicy` test tree. We'll get a conflict in any case and I'll solve it. Let's merge 
- https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/901

before this

zeeke commented on test/conformance/tests/test_no_policy.go:
```diff
@@ -0,0 +1,299 @@
+			)
+		})
+	})
+})
```
This case is handled  in 
- https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/897

but no such priority has been implemented yet. That PR handles these collisions by making the first object win.

for example, if a namespaced network is created, then the created netattachdef has the owner annotation. When the operatornamespaced network is created, it finds the target netattachdefs already owned by another object and the controller does nothing.

adrianchiris commented on pkg/webhook/validate_networks_test.go:
```diff
@@ -0,0 +1,47 @@
+	defer func(previous string) { vars.Namespace = previous }(vars.Namespace)
+	vars.Namespace = "operator-namespace"
+
+	validNetworkNamespaces := []controllers.NetworkCRInstance{
```
any chance to have a "[]tcase" (list of test case struct) which contains the test name, the NetworkCRInstance and a `shouldFail` boolean ? then the test would iterate over entries , run validateNetworkNamespace and compare to expected result ?

SchSeba reviewed (approved):
from my side LGTM.

to me it's not an hard request to have the object retreval in the handler, but it's nice to have.

zeeke commented:
I added the handleNetAttachDef function and it actually made the logic a little straightforward.

@adrianchiris @SchSeba please take a look

SchSeba reviewed (approved):
I link this new approach!

I learned something new :) 

LGTM

adrianchiris commented on controllers/generic_network_controller.go:
```diff
@@ -202,11 +210,36 @@ func (r *genericNetworkReconciler) SetupWithManager(mgr ctrl.Manager) error {
+func (r *genericNetworkReconciler) handleNetAttDef(ctx context.Context, obj client.Object) []reconcile.Request {
+	instance := r.controller.GetObject()
+	nadNamespacedName := types.NamespacedName{Namespace: obj.GetNamespace(), Name: obj.GetName()}
+	err := r.Get(ctx, nadNamespacedName, instance)
```
potentially we might have both existing and the other (from the sriov network operator namespace) is the actual owner.

maybe we should add both if they exist ?

zeeke commented on controllers/generic_network_controller.go:
```diff
@@ -163,6 +164,13 @@ func (r *genericNetworkReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 				return reconcile.Result{}, nil
 			}
 
+			if instance.GetNamespace() == netAttDef.Namespace {
```
This is the problem of multiple network objects referring the same NAD, though these changes makes the situation a little worse. 

I'm handling it here:
- https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/897

Is it ok to handle it in that subsequent PR?

zeeke commented:
> Hi @zeeke,
> 
> our QE did the following test on this PR
> 
> * Create sriovNetwork/net1 in ns1
> * Operator creates NAD/net1 in ns1
> * Update sriovNetwork/net1 with new resourceName
> * NAD/net1 is also updated (but no changes in spec as expected)
> 
> And on the last step when he check the NAD/net1 the owner reference was removed. So we probably have an issue with the update.

fixed.

@adrianchiris @e0ne @SchSeba please, take another look

adrianchiris commented on controllers/sriovnetwork_controller_test.go:
```diff
@@ -342,6 +342,68 @@ var _ = Describe("SriovNetwork Controller", Ordered, func() {
+
+				Eventually(func(g Gomega) {
+					netAttDef = &netattdefv1.NetworkAttachmentDefinition{}
+					err = util.WaitForNamespacedObject(netAttDef, k8sClient, "ns-blue", cr.GetName(), util.RetryInterval, util.Timeout)
```
nit: no need to call this wait func, object already exists, you can just call g.Expect(dynclient.Get(....)).To(Succeed())
its all wrapped in Eventually block anyway

adrianchiris commented on controllers/sriovnetwork_controller_test.go:
```diff
@@ -342,6 +342,68 @@ var _ = Describe("SriovNetwork Controller", Ordered, func() {
+
+				// Delete the SriovNetwork
+				err = k8sClient.Delete(ctx, &cr)
+				Expect(err).NotTo(HaveOccurred())
```
lets wait for the underlying net-attach-def to be deleted ?

adrianchiris commented on controllers/sriovnetwork_controller_test.go:
```diff
@@ -342,6 +342,68 @@ var _ = Describe("SriovNetwork Controller", Ordered, func() {
 				Should(Succeed())
 		})
+
+		Context("When the SriovNetwork namespace is not equal to the operator one", func() {
```
can you add a case where the user requests to create nad in a different ns than the current one ?
then check nad doesnt get created with a `Consistently` block ?

adrianchiris commented on controllers/generic_network_controller.go:
```diff
@@ -202,11 +215,41 @@ func (r *genericNetworkReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	instance := r.controller.GetObject()
+	nadNamespacedName := types.NamespacedName{Namespace: obj.GetNamespace(), Name: obj.GetName()}
+
+	log.Log.WithName("XXX").Info("handleNetAttDef", "nadNamespacedName", nadNamespacedName, "vars.Namespace", vars.Namespace)
```
nit: remove debug logs (XXX/XXX11 etc)

adrianchiris reviewed (commented):
final nits otherwise LGTM

zeeke commented on controllers/sriovnetwork_controller_test.go:
```diff
@@ -390,3 +475,12 @@ func generateExpectedNetConfig(cr *sriovnetworkv1.SriovNetwork) string {
+	ExpectWithOffset(1, err).NotTo(HaveOccurred())
+
+	k8sClient.DeleteAllOf(ctx, &netattdefv1.NetworkAttachmentDefinition{}, client.InNamespace(namespace))
+	ExpectWithOffset(1, err).NotTo(HaveOccurred())
```
sure, I put the delete and list calls in an eventually block

adrianchiris commented on controllers/sriovnetwork_controller_test.go:
```diff
@@ -390,3 +475,24 @@ func generateExpectedNetConfig(cr *sriovnetworkv1.SriovNetwork) string {
+	ctx := context.Background()
+	EventuallyWithOffset(1, func(g Gomega) {
+		err := k8sClient.DeleteAllOf(ctx, &sriovnetworkv1.SriovNetwork{}, client.InNamespace(namespace))
+		g.Expect(1, err).NotTo(HaveOccurred())
```
g.Expect(err) ? here and below

adrianchiris commented:
lgtm, lets have UT pass
